mars/fuse.pyx

```diff
@@ -96,30 +96,27 @@ class Fusion(object):
                 return n.composed[-1]
             return n
 
-        observed = set()
-
         composed_nodes = node.composed
         nodes_set = set(composed_nodes)
         tail_node = composed_nodes[-1]
         self._graph.add_node(tail_node)
-        observed.add(tail_node)
         q = deque()
         q.append(tail_node)
         while len(q) > 0:
             cur_node = q.pop()
             if not cur_node.inputs:
                 continue
-            for pre in cur_node.inputs:
+            is_first = cur_node is composed_nodes[0]
+            inputs = cur_node.inputs if not is_first else \
+                self._graph.predecessors(node)
+            for pre in inputs:
                 pre = get_node(pre)
                 if pre in nodes_set:
                     self._graph.add_node(pre)
                 if pre in self._graph:
                     self._graph.add_edge(pre, cur_node)
                 if pre in nodes_set:
                     q.appendleft(pre)
-                    observed.add(pre)
-        if len(observed) != len(nodes_set):
-            raise InvalidComposedNodeError("Invalid composed node data")
```

Comment:
The error definition can be removed when it is not needed.

Reply:
Removed already.